src/transaction/log_manager.c

```diff
@@ -10214,6 +10219,482 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	      LSA_COPY (start_lsa, &log_rec_header->forw_lsa);
+	      return NO_ERROR;
+	    }
+	  LOG_READ_ADD_ALIGN (thread_p, sizeof (*dummy), &process_lsa, log_page_p);
+	}
+
+      if (process_lsa.pageid != forw_lsa.pageid)
+	{
+	  if (LSA_ISNULL (&forw_lsa))
+	    {
+	      return ER_FAILED;
+	    }
+
+	  if (logpb_fetch_page (thread_p, &forw_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+	    {
+	      return ER_FAILED;
+	    }
+	}
+
+      LSA_COPY (&process_lsa, &forw_lsa);
+    }
+
+  return ER_FAILED;
+}
+
+/*consumer in other PR when PR related to THREAD handling is merged */
+int
+cdc_get_logitem_info (THREAD_ENTRY * thread_p, LOG_LSA * start_lsa, int *total_length, int *num_log_info)
+{
+  int rv;
+#if 0
+  struct timeval cur;
+  struct timespec timeout;
+  int status;
+
+  char *log_infos = NULL;
+  LOG_INFO_ENTRY *consume;
+
+  assert (!LSA_ISNULL (start_lsa));
+
+  *num_log_info = 0;
+  *total_length = 0;
+
+  gettimeofday (&cur, NULL);
+  timeout.tv_sec = cur.tv_sec + log_Reader_info.extraction_timeout;
+  timeout.tv_nsec = cur.tv_usec * 1000;
+
+  if (log_info_queue->is_empty ())
+    {
+      status = pthread_cond_timedwait (consume_cv, &mutex, timeout);
+      if (status == ETIMEDOUT)
+	{
+	  return ER_CDC_EXTRACTION_TIMEOUT;
+	}
+    }
+
+  while (log_info_queue->is_empty () == false && (*num_log_info < log_Reader_info.max_log_item))
+    {
+      /* *INDENT-OFF* */
+      log_info_queue->consume (consume);
+      /* *INDENT-ON* */
+      if (LSA_GE (&consume->start_lsa, start_lsa))
+	{
+	  log_infos = (char *) realloc (log_infos, *total_length + consume->length + MAX_ALIGNMENT);
+
+	  memcpy (PTR_ALIGN (log_infos + *total_length, MAX_ALIGNMENT), PTR_ALIGN (consume->log_info, MAX_ALIGNMENT),
+		  consume->length);
+
+	  *total_length =
+	    (PTR_ALIGN (log_infos + *total_length, MAX_ALIGNMENT) + consume->length) - PTR_ALIGN (log_infos,
+												  MAX_ALIGNMENT);
+
+	  (*num_log_info)++;
+
+	  LSA_COPY (start_lsa, &consume->start_lsa);
+
+	  if (consume->log_info != NULL)
+	    {
+	      free (consume->log_info);
+	    }
+
+	  if (consume != NULL)
+	    {
+	      free (consume);
+	    }
+	}
+
+      server_comm_buf.log_Infos = log_infos;
+      server_comm_buf.log_Infos_length = *total_length;
+    }
+#endif
+  return NO_ERROR;
+}
+
+int
+cdc_finalize ()
+{
+  int i = 0;
+
+  if (log_Reader_info.num_user > 0)
+    {
+      for (i = 0; i < log_Reader_info.num_user; i++)
+	{
+	  free (log_Reader_info.user[i]);
+	}
+    }
+  if (log_Reader_info.num_class > 0)
+    {
+      free (log_Reader_info.class_oids);
+    }
+
+for (auto iter:tran_users)
+    {
+      /*std::unordered_map<TRANID, char *> tran_users */
+      free (iter.second);
+    }
+
+  while (!log_info_queue->is_empty ())
+    {
+      LOG_INFO_ENTRY *tmp;
+      log_info_queue->consume (tmp);
+
+      if (tmp->log_info != NULL)
+	{
+	  free (tmp->log_info);
+	}
+
+      if (tmp != NULL)
+	{
+	  free (tmp);
+	}
+    }
+
+  delete log_info_queue;
```

Comment:
- global variable?
- How about using only new/delete or malloc/dealloc?

Reply:
* yes, log_info_queue is global variable
* log_info in LOG_INFO_ENTRY has memory using malloc, so I used free() (dealloc). and log_info_queue(lock free::circular_queue)  has been created using 'new'. Do you mean that uses 'new' instead of 'malloc' for consistency?